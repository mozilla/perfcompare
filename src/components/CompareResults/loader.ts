import { defer } from 'react-router-dom';

import { repoMap, frameworks } from '../../common/constants';
import {
  fetchCompareResults,
  fetchFakeCompareResults,
  fetchRecentRevisions,
} from '../../logic/treeherder';
import { Changeset, Repository } from '../../types/state';
import { FakeCommitHash, Framework } from '../../types/types';

// This function checks and sanitizes the input values, then returns values that
// we can then use in the rest of the application.
function checkValues({
  baseRev,
  baseRepo,
  newRevs,
  newRepos,
  framework,
}: {
  baseRev: string | null;
  baseRepo: Repository['name'] | null;
  newRevs: string[];
  newRepos: Repository['name'][];
  framework: string | number | null;
}): {
  baseRev: string;
  baseRepo: Repository['name'];
  newRevs: string[];
  newRepos: Repository['name'][];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
} {
  if (baseRev === null) {
    throw new Error('The parameter baseRev is missing.');
  }

  if (baseRepo === null) {
    throw new Error('The parameter baseRepo is missing.');
  }

  const validRepoValues = Object.values(repoMap);
  if (!validRepoValues.includes(baseRepo)) {
    throw new Error(
      `The parameter baseRepo "${baseRepo}" should be one of ${validRepoValues.join(
        ', ',
      )}.`,
    );
  }

  if (framework === null) {
    framework = 1; // default to talos so that manually typing the URL is easier
  }

  const frameworkId = +framework as Framework['id'];
  if (Number.isNaN(frameworkId)) {
    throw new Error(
      `The parameter framework should be a number, but it is "${framework}".`,
    );
  }
  const frameworkName = frameworks.find(
    (entry) => entry.id === frameworkId,
  )?.name;

  if (!frameworkName) {
    throw new Error(
      `The parameter framework isn't a valid value: "${framework}".`,
    );
  }
  if (!newRevs.length) {
    return {
      baseRev,
      baseRepo,
      newRevs: [baseRev],
      newRepos: [baseRepo],
      frameworkId,
      frameworkName,
    };
  }

  if (newRevs.length !== newRepos.length) {
    throw new Error(
      'There should be as many "newRepo" parameters as there are "newRev" parameters.',
    );
  }
  if (!newRepos.every((newRepo) => validRepoValues.includes(newRepo))) {
    throw new Error(
      `Every parameter newRepo "${newRepos.join(
        '", "',
      )}" should be one of ${validRepoValues.join(', ')}.`,
    );
  }

  return {
    baseRev,
    baseRepo,
    newRevs,
    newRepos,
    frameworkId,
    frameworkName,
  };
}

// This is essentially a glue to call the related function from
// /logic/treeherder.ts for all the revs we need results for.
async function fetchCompareResultsOnTreeherder({
  baseRev,
  baseRepo,
  newRevs,
  newRepos,
  framework,
}: {
  baseRev: string;
  baseRepo: Repository['name'];
  newRevs: string[];
  newRepos: Repository['name'][];
  framework: Framework['id'];
}) {
  const promises = newRevs.map((newRev, i) =>
    fetchCompareResults({
      baseRev,
      baseRepo,
      newRev,
      newRepo: newRepos[i],
      framework,
    }),
  );
  return Promise.all(promises);
}

const fakeCommitHashes: FakeCommitHash[] = [
  'bb6a5e451dace3b9c7be42d24c9272738d73e6db',
  '9d50665254899d8431813bdc04178e6006ce6d59',
  'a998c42399a8fcea623690bf65bef49de20535b4',
];

// This is essentially a glue to call the related function from
// /logic/treeherder.ts for all the revs we need results for.
async function fetchAllFakeCompareResults() {
  const promises = fakeCommitHashes.map((hash) =>
    fetchFakeCompareResults(hash),
  );
  return Promise.all(promises);
}

// This function is responsible for fetching the data from the URL. It's called
// by React Router DOM when the compare-results path is requested.
// It uses the URL parameters as inputs, and returns all the fetched data to the
// React components through React Router's useLoaderData hook.
export async function loader({ request }: { request: Request }) {
  const url = new URL(request.url);
  const useFakeData = url.searchParams.has('fakedata');
  if (useFakeData) {
    const results = await fetchAllFakeCompareResults();
    // They're all based on the same rev
    const baseRev = results[0][0].base_rev;
    // And the same repository
    const baseRepo = results[0][0].base_repository_name;

    const newRevs = fakeCommitHashes;
    const newRepos = results.map((result) => result[0].new_repository_name);
    const frameworkId = results[0][0].framework_id;
    const frameworkName =
      frameworks.find((entry) => entry.id === frameworkId)?.name ?? '';

    return {
      results,
      baseRev,
      baseRepo,
      newRevs,
      newRepos,
      frameworkId,
      frameworkName,
    };
  }

  const baseRevFromUrl = url.searchParams.get('baseRev');
  const baseRepoFromUrl = url.searchParams.get('baseRepo') as
    | Repository['name']
    | null;
  const newRevsFromUrl = url.searchParams.getAll('newRev');
  const newReposFromUrl = url.searchParams.getAll(
    'newRepo',
  ) as Repository['name'][];
  const frameworkFromUrl = url.searchParams.get('framework');

  const { baseRev, baseRepo, newRevs, newRepos, frameworkId, frameworkName } =
    checkValues({
      baseRev: baseRevFromUrl,
      baseRepo: baseRepoFromUrl,
      newRevs: newRevsFromUrl,
      newRepos: newReposFromUrl,
      framework: frameworkFromUrl,
    });

  const resultsPromise = fetchCompareResultsOnTreeherder({
    baseRev,
    baseRepo,
    newRevs,
    newRepos,
    framework: frameworkId,
  });

  // For each of these requests, we get a list of 1 item because we request one
  // specific hash.
  // TODO what happens if there's no result?
  const baseRevInfoPromise = fetchRecentRevisions({
    repository: baseRepo,
    hash: baseRev,
  }).then((listOfRevisions) => listOfRevisions[0]);
  const newRevsInfoPromises = newRevs.map((newRev, i) =>
    fetchRecentRevisions({ repository: newRepos[i], hash: newRev }).then(
      (listOfRevisions) => listOfRevisions[0],
    ),
  );

  const [baseRevInfo, ...newRevsInfo] = await Promise.all([
    baseRevInfoPromise,
    ...newRevsInfoPromises,
  ]);

  return defer({
    results: resultsPromise,
    baseRev,
    baseRevInfo,
    baseRepo,
    newRevs,
    newRevsInfo,
    newRepos,
    frameworkId,
    frameworkName,
  });
}

type DeferredLoaderData = {
  results: Promise<unknown>;
  baseRev: string;
  baseRevInfo: Changeset;
  baseRepo: Repository['name'];
  newRevs: string[];
  newRevsInfo: Changeset[];
  newRepos: Repository['name'][];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
};

//had to be more explicit with the type because the defer
//function returns a an inaccessible type
export type LoaderReturnValue = DeferredLoaderData;
