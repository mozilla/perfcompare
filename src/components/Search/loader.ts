import { repoMap, frameworks } from '../../common/constants';
import { memoizedFetchRevisionForRepository } from '../../logic/treeherder';
import { Changeset, Repository } from '../../types/state';
import { Framework } from '../../types/types';

const DEFAULT_VALUES = {
  newRev: null,
  newRevInfo: null,
  newRepo: 'try' as Repository['name'],
  frameworkId: 1 as Framework['id'],
  frameworkName: 'talos' as Framework['name'],
};

// This function checks and sanitizes the input values, then returns values that
// we can then use in the rest of the application.
function checkValues({
  newRev,
  newRepo,
  frameworkName,
}: {
  newRev: string | null;
  newRepo: Repository['name'] | null;
  frameworkName: Framework['name'] | null;
}): null | {
  newRev: string;
  newRepo: Repository['name'];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
} {
  if (newRev === null || newRepo === null) {
    return null;
  }

  const validRepoValues = Object.values(repoMap);
  if (!validRepoValues.includes(newRepo)) {
    console.warn(`The repository ${newRepo} wasn't found in our list.`);
    return null;
  }

  if (frameworkName === null) {
    frameworkName = DEFAULT_VALUES.frameworkName;
  }

  let frameworkId = frameworks.find(
    (entry) => entry.name === frameworkName,
  )?.id;

  if (frameworkId === undefined) {
    // Default to talos if the entry wasn't found.
    console.warn(
      `The framework entry for ${frameworkName} wasn't found, defaulting to talos.`,
    );
    frameworkId = DEFAULT_VALUES.frameworkId;
    frameworkName = DEFAULT_VALUES.frameworkName;
  }

  return {
    newRev,
    newRepo,
    frameworkId,
    frameworkName,
  };
}

// This function is responsible for fetching the data from the URL. It's called
// by React Router DOM when the compare-results path is requested.
// It uses the URL parameters as inputs, and returns all the fetched data to the
// React components through React Router's useLoaderData hook.
export async function loader({ request }: { request: Request }) {
  const url = new URL(request.url);
  const newRevFromUrl = url.searchParams.get('newRev');
  const newRepoFromUrl = url.searchParams.get('newRepo') as
    | Repository['name']
    | null;
  const frameworkFromUrl = url.searchParams.get('frameworkName') as
    | Framework['name']
    | null;

  const checkedValues = checkValues({
    newRev: newRevFromUrl,
    newRepo: newRepoFromUrl,
    frameworkName: frameworkFromUrl,
  });
  if (!checkedValues) {
    return DEFAULT_VALUES;
  }

  const { newRev, newRepo, frameworkId, frameworkName } = checkedValues;

  const newRevInfo = await memoizedFetchRevisionForRepository({
    repository: newRepo,
    hash: newRev,
  });

  if (!newRevInfo) {
    // The search returned no result.
    return DEFAULT_VALUES;
  }

  return {
    newRev,
    newRevInfo,
    newRepo,
    frameworkId,
    frameworkName,
  };
}

// Be explicit with the returned type to control it better than if we were
// inferring it.
export type LoaderReturnValue = {
  newRev: string;
  newRevInfo: Changeset;
  newRepo: Repository['name'];
  frameworkId: Framework['id'];
  frameworkName: Framework['name'];
};
