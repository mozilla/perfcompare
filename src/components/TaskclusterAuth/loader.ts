import {
  storeUserToken,
  storeUserCredentials,
} from '../../logic/credentials-storage';
import {
  retrieveTaskclusterUserCredentials,
  retrieveTaskclusterToken,
} from '../../logic/taskcluster';

export async function loader({ request }: { request: Request }) {
  const url = new URL(request.url);

  const taskclusterCode = url.searchParams.get('code') as string;
  const state = url.searchParams.get('state');

  const requestState = sessionStorage.getItem('requestState');

  const rootUrl = sessionStorage.getItem('taskclusterUrl') as string;

  if (!taskclusterCode) {
    throw new Error(`Error when getting Taskcluster code from URL.`);
  }

  if (!requestState || !state || requestState !== state) {
    throw new Error(
      `Error missing or not matching request state with URL state`,
    );
  }

  const tokenResponse = await retrieveTaskclusterToken(
    rootUrl,
    taskclusterCode,
  );

  storeUserToken(rootUrl, tokenResponse);

  // fetch access token with token Bearer
  const userCredentials = await retrieveTaskclusterUserCredentials(
    rootUrl,
    tokenResponse.access_token,
  );

  storeUserCredentials(rootUrl, userCredentials);

  return userCredentials;
}

export type LoaderReturnValue = Awaited<ReturnType<typeof loader>>;
