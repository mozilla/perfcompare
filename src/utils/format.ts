const dateFormatter = new Intl.DateTimeFormat('en-US', {
  dateStyle: 'medium',
});

export function formatDate(date: Date) {
  return dateFormatter.format(date);
}

export function formatDateRange(date1: Date, date2: Date) {
  return formatDate(date1) + ' - ' + formatDate(date2);
}
