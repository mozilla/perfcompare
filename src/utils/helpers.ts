import {
  frameworkMap,
  devToolsFramework,
  baseDocsURL,
  removedOldTestDevTools,
  nonDocumentedTestsDevTools,
  supportedPerfdocsFrameworks,
} from '../common/constants';
import type { Repository, Changeset } from '../types/state';
import type { Framework, SupportedPerfdocsFramework } from '../types/types';

const truncateHash = (revision: Changeset['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Changeset) => {
  const { repository_id: repositoryId, revisions } = item;
  const isTry = repositoryId === 4;
  const lastUsefulRevision =
    isTry && revisions.length > 1 ? revisions[1] : revisions[0];
  const lastUsefulSummary = lastUsefulRevision.comments.slice(
    0,
    lastUsefulRevision.comments.indexOf('\n'),
  );
  return lastUsefulSummary;
};

// ensure all numbers display two digits
const formatNumber = (number: number): string => {
  if (number < 10) {
    return `0${number}`;
  } else return String(number);
};

const formatDate = (timestamp: number): string => {
  const date = new Date(timestamp * 1000);
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = String(date.getFullYear()).slice(2);
  const hours = date.getHours();
  const minutes = date.getMinutes();

  return `${formatNumber(month)}/${formatNumber(day)}/${year} ${formatNumber(
    hours,
  )}:${formatNumber(minutes)}`;
};

const getTreeherderURL = (
  revision: Changeset['revision'],
  repository: Repository['name'],
) =>
  `https://treeherder.mozilla.org/jobs?repo=${repository}&revision=${revision}`;

const createDevtoolsDocsUrl = (
  supportedFramework: string,
  urlReadySuite: string,
  suite: string,
) => {
  let linkSupported = true;

  let devtoolsDocsURL = `${baseDocsURL}/devtools/tests/${supportedFramework}.html#${urlReadySuite}`;
  if (
    suite === removedOldTestDevTools ||
    nonDocumentedTestsDevTools.includes(suite)
  ) {
    devtoolsDocsURL = '';
    linkSupported = false;
  }
  return { devtoolsDocsURL, linkSupported };
};

const getDocsURL = (suite: string, framework_id: Framework['id']) => {
  const framework = frameworkMap[framework_id];
  const supportedFramework =
    supportedPerfdocsFrameworks[framework as SupportedPerfdocsFramework];
  const urlReadySuite = suite.replace(/:|\s|\.|_/g, '-').toLowerCase();

  let docsURL = '';
  let isLinkSupported = true;

  if (framework_id === devToolsFramework.id) {
    const { devtoolsDocsURL, linkSupported } = createDevtoolsDocsUrl(
      supportedFramework,
      urlReadySuite,
      suite,
    );

    isLinkSupported = linkSupported;
    docsURL = devtoolsDocsURL;
  } else if (supportedFramework) {
    docsURL = `${baseDocsURL}/testing/perfdocs/${supportedFramework}.html#${urlReadySuite}`;
  } else {
    isLinkSupported = false;
  }
  return { docsURL, isLinkSupported };
};

// TO DO: Review if this method is still needed
const setConfidenceClassName = (confidenceText: string | null) => {
  return confidenceText || 'unknown-confidence';
};

const swapArrayElements = <T>(
  array: T[],
  index1: number,
  index2: number,
): T[] => {
  if (
    index1 !== index2 &&
    [index1, index2].every((index) => index < array.length && index >= 0)
  ) {
    const newArray = [...array];
    [newArray[index1], newArray[index2]] = [newArray[index2], newArray[index1]];
    return newArray;
  }
  return array;
};

export {
  formatDate,
  getLatestCommitMessage,
  getTreeherderURL,
  setConfidenceClassName,
  swapArrayElements,
  truncateHash,
  getDocsURL,
};
