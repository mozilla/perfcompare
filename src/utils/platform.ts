import type { PlatformShortName } from '../types/state';

export const getPlatformShortName = (
  platformName: string,
): PlatformShortName => {
  if (platformName.toLowerCase().includes('linux')) return 'Linux';
  if (
    platformName.toLowerCase().includes('osx') ||
    platformName.toLowerCase().includes('os x')
  )
    return 'OSX';
  if (platformName.toLowerCase().includes('windows')) return 'Windows';
  if (platformName.toLowerCase().includes('android')) return 'Android';
  return 'Unspecified';
};
