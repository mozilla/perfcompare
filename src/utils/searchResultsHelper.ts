import type { Revision } from '../types/state';

const truncateHash = (revision: Revision['revision']) => revision.slice(0, 12);

// return only most recent commit message
// first commit is usually 'try_task_config'
const getLatestCommitMessage = (item: Revision) =>
  item.revisions[item.revisions.length - 1].comments;

export { truncateHash, getLatestCommitMessage };
