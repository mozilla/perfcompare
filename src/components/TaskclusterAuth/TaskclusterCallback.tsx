import { Suspense } from 'react';

import { Typography, Box, CircularProgress } from '@mui/material';
import { useLoaderData, Await } from 'react-router-dom';

import { LoaderReturnValue } from './loader';

function WaitingForCredentials() {
  return (
    <Box
      sx={{
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        gap: 3,
        margin: 3,
      }}
    >
      <Typography variant='h2' component='h2'>
        Retrieving Taskcluster credentials...
      </Typography>
      <CircularProgress />
    </Box>
  );
}

function CredentialsFound() {
  return (
    <>
      {/* This won't be shown because the window closes after the promise resolves,
          but is useful for tests. */}
      Credentials were found!
    </>
  );
}

export default function TaskclusterCallback() {
  const { retrievalPromise } = useLoaderData() as LoaderReturnValue;

  return (
    <>
      <Suspense fallback={<WaitingForCredentials />}>
        <Await resolve={retrievalPromise}>
          <CredentialsFound />
        </Await>
      </Suspense>
    </>
  );
}
