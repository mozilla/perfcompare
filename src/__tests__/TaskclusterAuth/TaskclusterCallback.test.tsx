import { loader } from '../../components/TaskclusterAuth/loader';
import TaskclusterCallback from '../../components/TaskclusterAuth/TaskclusterCallback';
import { getLocationOrigin } from '../../utils/location';
import {
  FetchMockSandbox,
  renderWithRouter,
  screen,
} from '../utils/test-utils';

jest.mock('../../utils/location');
const mockedGetLocationOrigin = getLocationOrigin as jest.Mock;

describe('Taskcluster Callback', () => {
  function setup({ inputState }: { inputState: string }) {
    // Make window.close a noop so that the component can be rendered after the
    // authentication process.
    jest.spyOn(window, 'close').mockImplementation(() => {});

    sessionStorage.setItem('requestState', inputState);
    sessionStorage.setItem(
      'taskclusterUrl',
      'https://firefox-ci-tc.services.mozilla.com',
    );

    mockedGetLocationOrigin.mockImplementation(() => 'http://localhost:3000');
  }

  it('should fetch credentials with token bearer', async () => {
    const inputCode = 'dwcygG5HQNaLiRe3RcTCbQ';
    const inputState = 'OkCrH5isZncYqeJbRDelN';
    const returnedBearerToken = 'RnVpOGJtdDZTb3FlWW5PVUxVclprQQ==';
    const returnedUserToken = 'jQWJVQdeRceT-YymPwTWagPh2PwJr0RmmZyL1uAfMSWg';
    const returnedClientId =
      'mozilla-auth0/ad|Mozilla-LDAP|ldapuser/perfcompare-localhost-3000-client-OCvzh5';

    setup({ inputState });

    (window.fetch as FetchMockSandbox).post(
      'https://firefox-ci-tc.services.mozilla.com/login/oauth/token',
      {
        access_token: returnedBearerToken,
        token_type: 'Bearer',
      },
    );

    (window.fetch as FetchMockSandbox).get(
      'begin:https://firefox-ci-tc.services.mozilla.com/login/oauth/credentials',
      {
        expires: '2024-05-20T14:07:40.828Z',
        credentials: {
          clientId: returnedClientId,
          accessToken: returnedUserToken,
        },
      },
    );
    renderWithRouter(<TaskclusterCallback />, {
      route: '/taskcluster-auth',
      search: `?code=${inputCode}&state=${inputState}`,
      loader,
    });

    expect(
      await screen.findByText(/Credentials were found/),
    ).toBeInTheDocument();

    expect(window.fetch).toHaveBeenCalledWith(
      'https://firefox-ci-tc.services.mozilla.com/login/oauth/token',
      {
        method: 'POST',
        body: expect.any(URLSearchParams) as URLSearchParams,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      },
    );
    const requestBody = Object.fromEntries(
      (window.fetch as jest.Mock).mock.calls[0][1].body as URLSearchParams,
    );

    expect(requestBody).toEqual({
      client_id: 'perfcompare-localhost-3000-client',
      code: inputCode,
      grant_type: 'authorization_code',
      redirect_uri: 'http://localhost/taskcluster-auth',
    });

    expect(window.fetch).toHaveBeenLastCalledWith(
      'https://firefox-ci-tc.services.mozilla.com/login/oauth/credentials',
      {
        headers: {
          Authorization: `Bearer ${returnedBearerToken}`,
          'Content-Type': 'aplication/json',
        },
      },
    );

    expect(localStorage.userTokens).toBe(
      `{"https://firefox-ci-tc.services.mozilla.com":{"access_token":"${returnedBearerToken}","token_type":"Bearer"}}`,
    );

    expect(localStorage.userCredentials).toBe(
      `{"https://firefox-ci-tc.services.mozilla.com":{"expires":"2024-05-20T14:07:40.828Z","credentials":{"clientId":"${returnedClientId}","accessToken":"${returnedUserToken}"}}}`,
    );

    expect(window.close).toHaveBeenCalled();
  });

  it('should show a spinner while waiting for the credentials', async () => {
    const inputCode = 'RANDOM_CODE';
    const inputState = 'RANDOM_STATE';
    setup({ inputState });

    const neverResolvedPromise = new Promise(() => {});

    (window.fetch as FetchMockSandbox).post(
      'https://firefox-ci-tc.services.mozilla.com/login/oauth/token',
      neverResolvedPromise,
    );

    (window.fetch as FetchMockSandbox).get(
      'begin:https://firefox-ci-tc.services.mozilla.com/login/oauth/credentials',
      neverResolvedPromise,
    );
    renderWithRouter(<TaskclusterCallback />, {
      route: '/taskcluster-auth',
      search: `?code=${inputCode}&state=${inputState}`,
      loader,
    });

    expect(
      await screen.findByText(/Retrieving Taskcluster credentials.../),
    ).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it('should show an error if it fails to fetch', async () => {
    // Because this test will throw an error, a lot of errors will be output to
    // the console. Let's silence them so that the test output stays clean.
    jest.spyOn(console, 'error').mockImplementation();

    const inputCode = 'RANDOM_CODE';
    const inputState = 'RANDOM_STATE';

    setup({ inputState });

    (window.fetch as FetchMockSandbox).post(
      'https://firefox-ci-tc.services.mozilla.com/login/oauth/token',
      {
        status: 403,
      },
    );

    renderWithRouter(<TaskclusterCallback />, {
      route: '/taskcluster-auth',
      search: `?code=${inputCode}&state=${inputState}`,
      loader,
    });

    expect(
      await screen.findByRole('heading', {
        name: /Error when requesting Taskcluster: \(403\) Forbidden/,
      }),
    ).toBeInTheDocument();
  });
});
