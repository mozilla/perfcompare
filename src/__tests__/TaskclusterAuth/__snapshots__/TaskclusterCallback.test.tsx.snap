// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taskcluster Callback should show a spinner while waiting for the credentials 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-vmdzdb"
    >
      <h2
        class="MuiTypography-root MuiTypography-h2 css-15umudq-MuiTypography-root"
      >
        Retrieving Taskcluster credentials...
      </h2>
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-zhs9ko-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
    </div>
  </div>
</body>
`;
